src/lib.rs

```diff
@@ -197,66 +207,76 @@ impl<'a> Context<gimli::EndianBuf<'a, gimli::RunTimeEndian>> {
     }
 }
 
-impl<R: gimli::Reader> ResUnit<R> {
-    fn parse_lines(&mut self, sections: &DebugSections<R>) -> Result<(), Error> {
-        if self.lines.is_some() {
-            return Ok(());
+impl<R> ResUnit<R>
+where
+    R: gimli::Reader + Sync,
+    R::Offset: Sync,
+{
+    fn parse_lines(&self, sections: &DebugSections<R>) -> Result<&Lines<R>, Error> {
+        match *self.lines.get_or_create(|| {
+            let ilnp = sections.debug_line.program(
+                self.line_offset,
+                self.dw_unit.address_size(),
+                self.comp_dir.clone(),
+                self.comp_name.clone(),
+            )?;
+
+            let (lnp, mut sequences) = ilnp.sequences()?;
+            sequences.retain(|x| x.start != 0);
+            sequences.sort_by_key(|x| x.start);
+            Ok(Lines { lnp, sequences })
+        }) {
+            Ok(ref lines) => Ok(lines),
+            Err(e) => Err(e),
```

Comment:
Took me a while to figure out what this does. Perhaps `.as_ref().map_err(Error::clone)` is clearer?

Reply:
Yeah that looks better, thanks.